### ISSUE TITLE
Renaming a document?

### ISSUE THREAD

**Issue by bjornhellander** [dotnet/roslyn#69125]
Hi!

This is a follow-up on #6261. Tagging @sharwell for information.

I have made an attempt at using Renamer.RenameDocumentAsync (https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.rename.renamer.renamedocumentasync?view=roslyn-dotnet-4.6.0#microsoft-codeanalysis-rename-renamer-renamedocumentasync(microsoft-codeanalysis-document-microsoft-codeanalysis-rename-documentrenameoptions-system-string-system-collections-generic-ireadonlylist((system-string))-system-threading-cancellationtoken)) to rename a document in a StyleCop.Analyzers code fix, but I am having problems with symbols being renamed as well. When reading the documentation properly, this seems to be intended behaviour, but unfortunately not what I am looking for.

Before discarding my work, I would just like to check with you if you agree on this, i.e. that RenameDocumentAsync is not my friend if I **only** want to rename the document. I have not investigated properly, but I guess that RenameDocumentAsync also renames a contained type **if** the document name matches that type's name?

I also saw that Solution has a WithDocumentName method since a number of years. I have attempted an implementation using that method instead and it seems to work in the easy cases at least, i.e. without linked documents. Does anyone have any insights regarding linked documents as mentioned in #6261, before I spend more time on this method?

Or generally, what is the proper way to rename a file?

Tested with 4.4.0 and 4.6.0.

**Comment by sharwell** [comment]
`Renamer.RenameDocumentAsync` is the expected API. This API only _renames_ a document; it must not make changes to the content of the document.

**Comment by sharwell** [comment]
Edit: looks like `RenameDocumentAsync` does attempt to make changes to the content of the document. That... isn't helpful.

**Comment by sharwell** [comment]
Duplicate of #6261

**Comment by CyrusNajmabadi** [comment]
@ryzngard For more info.  My recollection is that the API was supposed to give you the individual steps you could perform to do the rename.  This might include renaming the symbol or not so that the consumer could decide if they wanted to do that.  For example, that's how the project system can prompt and then not do symbolic rename (or namespace moving) if they don't want that.

**Comment by sharwell** [comment]
Ah, so you can call `RenameDocumentActionSet.UpdateSolutionAsync` and pass an empty collection for `actions` and it will only rename the document itself. However, there is no way to bypass all the computation work involved in the symbol rename (it performs the semantic work as part of `RenameDocumentAsync` itself).

**Comment by ryzngard** [comment]
Yea, the intent was that if the symbol already matched document name, then we want to mantain that. There might be a better way to expose just renaming a document in a code fix separately

**Comment by bjornhellander** [comment]
Thanks for the input! I remember seeing the overload with the array the first time I looked at the Renamer class, but the documentation didn't help me understand it so I ignored that one. This time I think I just missed it. Anyway, using that version with an empty array worked fine. 👍 Seems like a very UI oriented API because of the actions with displayable descriptions, but I will continue investigating. Thanks again!
